### DISCUSSION TITLE
Warnings when compiled with GCC -pedantic option

### DISCUSSION THREAD

**Discussion by Nikitarc** [hathach/tinyusb#1441] [General]
When I compile my code base TinyUSB generates a lot of warnings because of the -pedantic (strict ISO C) option of GCC that I use. For example:

`../TinyUSB/src/common/tusb_types.h:457:4: warning: ISO C99 doesn't support unnamed structs/unions [-Wpedantic]`
`../TinyUSB/src/common/tusb_common.h:55:31: warning: ISO C restricts enumerator values to range of 'int' [-Wpedantic]`
`..\TinyUSB\src\host\usbh.c:1636: warning: ISO C forbids an empty translation unit [-Wpedantic]
`

It is easy to eliminate messages relating to "empty translation unit". The common solution is to add
`typedef int make_iso_compilers_happy;`
in the file tusb_options.h.

The message relating to "enumerator values" is present 1 time.
In the hid.h#L241, replace:
`GAMEPAD_BUTTON_31 = TU_BIT(31),`
by:
`GAMEPAD_BUTTON_31 = (int) TU_BIT(31),`

For the third type of warning it's more complicated: TinyUSB uses unnamed unions, which is not allowed by ISO C. 5 out of 10 of these unions are in dwc2_types.h.

Is it possible that one day TinyUSB will be modified to eliminate these warnings?

**Comment by hathach** [comment]
```
It is easy to eliminate messages relating to "empty translation unit". The common solution is to add
typedef int make_iso_compilers_happy;
in the file tusb_options.h.
```

this sound great, would you mind making an PR for this

```
GAMEPAD_BUTTON_31 = (int) TU_BIT(31),
```

this could cause an issue, since it becomes negative number, it is better to just drop this BUTTON_31, it is a bit mask anyway.

```
For the third type of warning it's more complicated: TinyUSB uses unnamed unions, which is not allowed by ISO C. 5 out of 10 of these unions are in dwc2_types.h.
```

unnamed struct/union is part of C11, you could change the std to c11 and it would be all good. We may eventually move to c11 from c99 (linux kernel now compile with c11 as well).

**Reply by Nikitarc** [reply]
The message with C11 is:
`../TinyUSB/src/common/tusb_common.h:55:31: warning: ISO C restricts enumerator values to range of 'int' [-Wpedantic]`
`   55 | #define TU_BIT(n)             (1UL << (n))`
`      |                               ^`
`../TinyUSB/src/class/hid/hid.h:241:23: note: in expansion of macro 'TU_BIT'`
`  241 |   GAMEPAD_BUTTON_31 = TU_BIT(31),`

=> in C99 or C11 an enum is an int. So GAMEPAD_BUTTON_31 is an int: there is an implicit conversion from UINT to INT. I only propose to make an explicit cast to remove the warning message.
This will not cause an issue because the current code already uses it as an int, even if used as a mask
 
Ok for C11 and the unamed union!

**Reply by hathach** [reply]
thanks for testing it out, I will change the gamepad to macro instead of enum then
